### ISSUE TITLE
Incorrect handling of global phase in controlled gate

### ISSUE THREAD

**Issue by jlapeyre** [Qiskit/qiskit-terra#5587]
### Information

qiskit-terra version: master

### What is the current behavior?

`gate.control()` where `gate` is a U3 gate may result in an incorrect controlled gate.

### Steps to reproduce the problem

This creates a matrix:

```python
def simplecirc():
    u3 = U3Gate(0.6531640181825153, 0.8265458760929685, 3.9681385296827623)
    mat = u3.to_matrix()
    qc = QuantumCircuit(1)
    qc.append(u3, [0])
    qc.global_phase = -2.3973422028878653
    return qc

def getcont():
    circ = simplecirc()
    pcirc = circ.power(2)
    ccirc = pcirc.control()
    gc = ccirc.data[0][0].definition
    return gc

Operator(getcont())
```

You can check that this matrix is incorrect. This bug was introduced in #5342 . In particular in the changes in qiskit/circuit/add_control.py. These changes only affect how the global phase in the input gate is handled.

This is perhaps not quite a MWE. But, omitting the method call `circ.power(2)` does *not* trigger the bug, that is the matrix is correct in this case.

**Comment by Cryoris** [comment]
I think this might also be fixed by #5588, at least when I check locally the matrix produced by `Operator(getcont())` matches the one from `Operator(simplecirc()).data ** 2`.

**Comment by ewinston** [comment]
This is also fixed by reverting the change @jlapeyre suggested.

**Comment by jlapeyre** [comment]
Yes. As discussed out of band with @Cryoris, and in agreement with the two comments above, I find that either reverting the change mentioned here, or making the change in the new PR from @Cryoris fixes this issue. But, of course, only either or, not both. There is another closely related bug that I am writing up. We may need to consider them together.

**Comment by Cryoris** [comment]
Both our issues are around `power/repeat` though right? If #5342 fixed some other issues I'm not sure we should revert it. @jlapeyre did you also find problems that are not related to `power/repeat`?

**Comment by jlapeyre** [comment]
Exactly, I presume there is a test in  #5342 specifically for the file that implements `control`, but I did not check. Anyway (as we discussed!) the global phase in repeat is wrong on it's own and needs to be fixed. It was probably two bugs, one in `control` and one in `repeat` that cancelled one another.

**Comment by jlapeyre** [comment]
Fixed by #5588
